clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java

```diff
@@ -461,6 +463,12 @@ private static int parseAcks(String acksString) {
                         " specified in value.serializer");
             }
 
+            if(checkExpectedOffsets && record.expectedOffset() == -1L) {
```

Comment:
to be clear, it's definitely possible to make the server decide message-by-message whether or not to append, but each approach I saw is more complicated and harder to maintain, and I think this client-side approach will be received more warmly by the community.